File: Java/src/main/java/com/example/tree/AvlTree.java
package com.example.tree;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class AvlNode {
    AvlNode left;
    AvlNode right;
    int height;
    int data;

    public AvlNode(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
        this.height = 0;
    }

}

public class AvlTree {
    private AvlNode root;

    public AvlTree() {
        this.root = null;
    }

    private int getHeight(AvlNode node) {
        if (node == null) {
            return -1;
        }
        return node.height;
    }

    private int getMaxHeight(AvlNode node) {
        if (node == null) {
            return -1;
        }
        return Math.max(getHeight(node.left), getHeight(node.right));
    }

    // 左旋
    private AvlNode leftRotate(AvlNode node) {
        if (node == null) {
            return null;
        }
        AvlNode newRoot = node.right;
        node.right = newRoot.left;
        newRoot.left = node;
        node.height = getMaxHeight(node) + 1;
        newRoot.height = getMaxHeight(newRoot) + 1;
        return newRoot;
    }

    // 右旋
    private AvlNode rightRotate(AvlNode node) {
        AvlNode newRoot = node.left;
        node.left = newRoot.right;
        newRoot.right = node;
        node.height = getMaxHeight(node) + 1;
        newRoot.height = getMaxHeight(newRoot) + 1;
        return newRoot;
    }

    // 左右旋
    private AvlNode leftRightRotate(AvlNode node) {
        node.left = leftRotate(node.left);
        return rightRotate(node);
    }

    // 右左旋
    private AvlNode rightLeftRotate(AvlNode node) {
        node.right = rightRotate(node.right);
        return leftRotate(node);
    }

    private AvlNode rebalance(AvlNode node, int data) {
        if (node != null) {
            int leftHeight = getHeight(node.left);
            int rightHeight = getHeight(node.right);

            if (rightHeight - leftHeight == 2) {
                if (data > node.right.data) {
                    node = leftRotate(node);
                } else {
                    node = rightLeftRotate(node);
                }

            } else if (leftHeight - rightHeight == 2) {
                if (data < node.left.data) {
                    node = rightRotate(node);
                } else {
                    node = leftRightRotate(node);
                }
            }
        }

        return node;
    }

    // 插入

    public void insert(int data) {
        root = insert(root, data);
    }

    private AvlNode insert(AvlNode node, int data) {
        if (node == null) {
            return new AvlNode(data);
        }
        if (data < node.data) {
            node.left = insert(node.left, data);
        } else if (data > node.data) {
            node.right = insert(node.right, data);
        }
        node = rebalance(node, data);
        node.height = getMaxHeight(node) + 1;
        return node;
    }

    public void delete(int data) {
        root = delete(root, data);
    }


    private AvlNode delete(AvlNode node, int data) {
        if (node == null) {
            return null;
        }
        if (data < node.data) {
            node.left = delete(node.left, data);
        } else if (data > node.data) {
            node.right = delete(node.right, data);
        } else {
            if (node.left == null && node.right == null) {
                node = null;
            } else if (node.left == null) {
                node = node.right;
            } else if (node.right == null) {
                node = node.left;
            } else {
                AvlNode miNode = findMin(node.right);
                node.data = miNode.data;
                node.right = delete(node.right, miNode.data);
            }
            if (node != null) {
                node = rebalance(node, data);
                node.height = getMaxHeight(node) + 1;
            }

        }
        return node;
    }

    private AvlNode findMin(AvlNode node) {
        if (node == null) {
            return null;
        }
        while (node.left != null) {
            node = node.left;
        }
        return node;
    }

    public List<Integer> inOrder() {
        List<Integer> list = new ArrayList<>();

        return inOrder(root, list);

    }

    private List<Integer> inOrder(AvlNode node, List<Integer> list) {
        if (node != null) {
            inOrder(node.left, list);
            list.add(node.data);
            inOrder(node.right, list);
        }
        return list;
    }

    public List<Integer> levelOrder() {
        List<Integer> list = new ArrayList<>();
        Queue<AvlNode> queue = new LinkedList<>();
        if (root != null) {
            queue.add(root);
            while (!queue.isEmpty()) {
                AvlNode node = queue.poll();
                list.add(node.data);
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
        }
        return list;

    }

    public boolean isEmpty() {
        return root == null;
    }
}

File: typescript/tree/avlTree/avlTree.ts


class AVLTreeNode<T> {
    value: T
    left: AVLTreeNode<T> | null = null
    right: AVLTreeNode<T> | null = null
    height: number = 0
    constructor(value: T) {
        this.value = value
        this.height = 1
    }
}

export class AVLTree<T = number> {
    root: AVLTreeNode<T> | null = null
    constructor() {
    }

    getMaxHeight(node: AVLTreeNode<T> | null): number {
        if (node === null) {
            return 0
        }
        return Math.max(this.getHeight(node.left), this.getHeight(node.right)) + 1
    }

    getHeight(node: AVLTreeNode<T> | null): number {
        if (node === null) {
            return 0
        }
        return node.height
    }

    // 左旋转
    leftRotate(node: AVLTreeNode<T>) {
        const newRoot = node.right
        node!.right = newRoot?.left!
        newRoot!.left = node
        node.height = this.getMaxHeight(node)
        newRoot!.height = this.getMaxHeight(newRoot)
        return newRoot
    }
    // 右旋转
    rightRotate(node: AVLTreeNode<T>) {
        const newRoot = node.left
        node.left = newRoot?.right!
        newRoot!.right = node
        node.height = this.getMaxHeight(node)
        newRoot!.height = this.getMaxHeight(newRoot)
        return newRoot
    }

    // 左右旋转
    leftRightRotate(node: AVLTreeNode<T>) {
        node.left = this.leftRotate(node.left!)
        return this.rightRotate(node)
    }

    // 右左旋转
    rightLeftRotate(node: AVLTreeNode<T>) {
        node.right = this.rightRotate(node.right!)
        return this.leftRotate(node)
    }

    rebalance(node: AVLTreeNode<T> | null, data: T) {
        if (node === null) {
            return null
        }
        if (this.getHeight(node.left) - this.getHeight(node.right) > 1) {
            if (this.getHeight(node.left?.left!) > this.getHeight(node.left?.right!)) {
                node = this.rightRotate(node)
            } else {
                node = this.leftRightRotate(node)
            }
        } else if (this.getHeight(node.right) - this.getHeight(node.left) > 1) {
            if (this.getHeight(node.right?.right!) > this.getHeight(node.right?.left!)) {
                node = this.leftRotate(node)
            } else {
                node = this.rightLeftRotate(node)
            }
        }
        return node
    }

    insert(value: T) {
        this.root = this.insertNode(this.root, value)
    }
    insertNode(root: AVLTreeNode<T> | null, value: T): AVLTreeNode<T> | null {
        if (root === null) {
            return new AVLTreeNode<T>(value)
        }
        if (value < root.value) {
            root.left = this.insertNode(root.left, value)
        } else if (value > root.value) {
            root.right = this.insertNode(root.right, value)
        }
        root!.height = this.getMaxHeight(root)
        root = this.rebalance(root, value)
        return root
    }

    delete(value: T) {
        this.root = this.deleteNode(this.root, value)
    }
    deleteNode(root: AVLTreeNode<T> | null, value: T): AVLTreeNode<T> | null {
        if (root === null) {
            return null
        }
        if (value < root.value) {
            root.left = this.deleteNode(root.left, value)
        } else if (value > root.value) {
            root.right = this.deleteNode(root.right, value)
        } else {
            if (!root.left && !root.right) {
                root = null
            } else if (!root.left) {
                root = root.right
            } else if (!root.right) {
                root = root.left
            } else {
                const minNode = this.findMin(root.right)!
                root.value = minNode.value // 将节点的值替换为右子树的最小值
                root.right = this.deleteNode(root.right, minNode.value) // 删除右子树的最小值
            }
        }
        if (root) {
            root.height = this.getMaxHeight(root)
            root = this.rebalance(root, value)
        }
        return root
    }

    findMin(node: AVLTreeNode<T> | null): AVLTreeNode<T> | null {
        if (node === null) {
            return null
        }
        while (node.left) {
            node = node.left
        }
        return node
    }

    findMax(node: AVLTreeNode<T> | null): AVLTreeNode<T> | null {
        if (node === null) {
            return null
        }
        while (node.right) {
            node = node.right
        }
        return node
    }

    levelOrder(): T[] {
        const result: T[] = []
        if (!this.root) {
            return result
        }
        const queue: AVLTreeNode<T>[] = []
        queue.push(this.root)
        while (queue.length) {
            const node = queue.shift()!
            result.push(node.value)
            if (node.left) {
                queue.push(node.left)
            }
            if (node.right) {
                queue.push(node.right)
            }
        }
        return result
    }

    isEmpty() {
        return this.root === null
    }
}
